sklearn/cluster/_k_means.pyx

```diff
@@ -8,167 +8,207 @@
 #
 # License: BSD 3 clause
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+
+np.import_array()
 
-from sklearn.utils.sparsefuncs_fast import assign_rows_csr
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
-cdef extern from "cblas.h":
-    double ddot "cblas_ddot"(int N, double *X, int incX, double *Y, int incY)
-    float sdot "cblas_sdot"(int N, float *X, int incX, float *Y, int incY)
 
-np.import_array()
+cpdef floating _inertia_dense(np.ndarray[floating, ndim=2, mode='c'] X,
+                              floating[::1] sample_weight, 
+                              floating[:, ::1] centers,
+                              int[::1] labels):
+    """Compute inertia for dense input data
+    
+    Sum of squared distance between each sample and it's assigned center.
+    """
+    cdef:
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int i, j, k
+        floating tmp, sample_inertia
+
+        floating inertia = 0.0
+
+    for i in range(n_samples):
+        j = labels[i]
+        sample_inertia = 0.0
+        for k in range(n_features):
+            tmp = X[i, k] - centers[j, k]
+            sample_inertia += tmp * tmp
+        inertia += sample_inertia * sample_weight[i]
 
+    return inertia
 
-@cython.boundscheck(False)
-@cython.wraparound(False)
-@cython.cdivision(True)
-cpdef DOUBLE _assign_labels_array(np.ndarray[floating, ndim=2] X,
-                                  np.ndarray[floating, ndim=1] sample_weight,
-                                  np.ndarray[floating, ndim=1] x_squared_norms,
-                                  np.ndarray[floating, ndim=2] centers,
-                                  np.ndarray[INT, ndim=1] labels,
-                                  np.ndarray[floating, ndim=1] distances):
-    """Compute label assignment and inertia for a dense array
 
-    Return the inertia (sum of squared distances to the centers).
+cpdef floating _inertia_sparse(X,
+                               floating[::1] sample_weight,
```

Comment:
This line seems to be off by one character.

I have been using this custom flake8 config for Cython code (https://github.com/apache/arrow/blob/master/python/.flake8.cython), you can run it with

```
flake8 --config flake8.cython 
```

EDIT: it's not off, it's github. I would still suggest pep8ing as much as you can, especially the long lines ;)

Reply:
I've been trying to respect flake8 most of the time, but I find that some line breaks cause less readability. But I agree that for github it's bad... I'll fix that